python/test/unit/spacewalk/satellite_tools/test_yum_src.py

```diff
@@ -213,6 +214,66 @@ def test_mirror_list_arch(self):
                 "http://host3/base/arch1/os/",
             ])
 
+    @patch("spacewalk.satellite_tools.repo_plugins.yum_src.initCFG", Mock())
+    @patch("spacewalk.satellite_tools.repo_plugins.yum_src.os.unlink", Mock())
+    @patch("urlgrabber.grabber.PyCurlFileObject", Mock())
+    @patch("spacewalk.common.rhnLog", Mock())
+    @patch("spacewalk.satellite_tools.repo_plugins.yum_src.fileutils.makedirs", Mock())
+    @patch("spacewalk.satellite_tools.repo_plugins.yum_src.etree.parse", MagicMock(side_effect=Exception))
+    def test_proxy_usage_with_mirrorlist(self):
+        cs = self._make_dummy_cs()
+        proxy_url = "http://proxy.example.com:8080"
+        proxy_user = "user"
+        proxy_pass = "pass"
+        cs.proxy_hostname = proxy_url
+        cs.proxy_user = proxy_user
+        cs.proxy_pass = proxy_pass
+        expected_url_list = []
+        url_list = [
+            "http://host1/base/arch1/os/",
+            "http://host2/",
+            "http://host3.com/",
+            ]
+        for url in url_list:
+            expected_url_list.append("{}?proxy={}&proxyuser={}&proxypass={}".format(url,
+                                                                                    quote(proxy_url),
+                                                                                    proxy_user,
+                                                                                    proxy_pass
+                                                                                    ))
+        grabber_mock = Mock()
+
+        with patch("spacewalk.satellite_tools.repo_plugins.yum_src.urlgrabber.urlgrab", grabber_mock):
+            with open(os.path.join(self.repo.root, "mirrorlist.txt"), "w") as fake_list:
+                fake_list.writelines([
+                    "http://host1/base/arch1/os/\n",
+                    "http://host2/\n",
+                    "http://host3.com/\n",
+                ])
+            mirrors = cs._get_mirror_list(self.repo, "https://fake/repo/url")
+
+            self.assertEqual(mirrors, expected_url_list)
+
+    def test_prep_zypp_repo_url_with_proxy(self):
+        cs = self._make_dummy_cs()
+        url_list = ["http://example.com/", "https://example.com/", "uln://example.com/", "uln:///channel"]
+        uln_list = [False, False, True, True]
+        proxy_url = "http://proxy.example.com:8080"
+        proxy_user = "user"
+        proxy_pass = "pass"
+        cs.proxy_hostname = proxy_url
+        cs.proxy_user = proxy_user
+        cs.proxy_pass = proxy_pass
+
+        for i in range(0, len(url_list)):
+            if not uln_list[i]:
+                expected_url = "{}?proxy={}&proxyuser={}&proxypass={}".format(url_list[i], quote(proxy_url), proxy_user, proxy_pass)
```

Comment:
Not sure here about the way of checking ULN. I supposed ULN just must have some query parameters, that's why it was set with `&`, but not `?`. Am I wrong?

Reply:
No, you are right, but this is the negative branch, so it should be right

Reply:
Or do you mean I shouldn't decide if a url is uln manually?